pkg/source/pulsar.go

```diff
@@ -286,9 +286,56 @@ func newAckTracker(size uint) *ackTracker {
 }
 
 func (t *ackTracker) ack(batchID int) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
 	if batchID < 0 {
 		return true
 	}
 	t.batchIDs.Clear(uint(batchID))
 	return t.batchIDs.None()
 }
+
+type ackTrackers struct {
+	mu       sync.RWMutex
+	trackers map[string]*ackTracker
+}
+
+func newAckTrackers(size int) *ackTrackers {
+	return &ackTrackers{
+		trackers: make(map[string]*ackTracker, size),
+	}
+}
+
+func (a *ackTrackers) key(msg pulsar.MessageID) string {
+	return fmt.Sprintf("%d:%d", msg.LedgerID(), msg.EntryID())
+}
+
+func (a *ackTrackers) tryAdd(msg pulsar.MessageID) (ok bool) {
+	key := a.key(msg)
+	a.mu.RLock()
+	_, ok = a.trackers[key]
+	a.mu.RUnlock()
+	if !ok {
+		a.mu.Lock()
+		a.trackers[key] = newAckTracker(uint(msg.BatchSize()))
```

Comment:
`_, ok = a.trackers[key]` condition should be rechecked after acquiring the lock.

Reply:
indeed

Reply:
I think maybe we could use sync.Map to handle this situation.
In our case, we have a read-heavy and write-light scenario. Each batch message involves only one write operation and a final delete key action. There is no need to modify the value.